dqvm/src/global-addresses.lisp

```diff
@@ -0,0 +1,181 @@
+   (permutation
+    :reader permutation
+    :writer update-permutation
+    :initarg :permutation
+    :type permutation
+    :documentation "Last qubit permutation evaluated, stored in a format suitable for use by APPLY-QUBIT-PERMUTATION.")
+
+   ;; The following attributes are calculated during instantiation.
+   (blocks-per-process
+    :reader blocks-per-process
+    :type alexandria:non-negative-fixnum
+    :documentation "Regular blocks per process (i.e., without including a possibly remaining block).")
+   (remainder-blocks
+    :reader remainder-blocks
+    :type alexandria:non-negative-fixnum
+    :documentation "Total number of remainder blocks.")
+   (number-of-blocks
+    :reader number-of-blocks            ; XXX remove the get- prefix
+    :type alexandria:non-negative-fixnum
+    :documentation "Total number of blocks among all address tables."))
+
+  (:default-initargs
+   :number-of-processes (mpi-comm-size)
+   :block-size *default-block-size*
+   :permutation nil)
+  (:documentation "Global address table."))
+
+(defmethod initialize-instance :after ((global-addresses global-addresses) &rest initargs)
+  (declare (ignore initargs))
+
+  (assert (not (minusp (number-of-qubits global-addresses))))
+  (assert (plusp (number-of-processes global-addresses)))
+  (assert (evenp (block-size global-addresses)))
+
+  (multiple-value-bind (bpp rem) (%blocks-per-process global-addresses)
+    (setf (slot-value global-addresses 'number-of-blocks) (%number-of-blocks global-addresses)
+          (slot-value global-addresses 'blocks-per-process) bpp
+          (slot-value global-addresses 'remainder-blocks) rem)))
+
+(defmethod print-object ((global-addresses global-addresses) stream)
+  (let ((*print-readably* nil)
+        (*print-pretty* nil))
+
+    (print-unreadable-object (global-addresses stream :type t :identity t)
+
+      (format stream "~{~A ~A~^ ~}"
+              (list (prin1-to-string :number-of-qubits) (number-of-qubits global-addresses)
+                    (prin1-to-string :number-of-processes) (number-of-processes global-addresses)
+                    (prin1-to-string :block-size) (block-size global-addresses)
+                    (prin1-to-string :permutation) (permutation global-addresses))))))
+
+(defmethod copy ((global-addresses global-addresses))
+  (let ((number-of-qubits (number-of-qubits global-addresses))
+        (number-of-processes (number-of-processes global-addresses))
+        (block-size (block-size global-addresses))
+        (permutation (permutation global-addresses)))
+
+    (make-instance 'global-addresses :number-of-qubits number-of-qubits
+                                     :number-of-processes number-of-processes
+                                     :block-size block-size
+                                     :permutation permutation)))
+
+(defmethod equal-p ((g1 global-addresses) (g2 global-addresses))
+  (let ((number-of-qubits1 (number-of-qubits))
```

Comment:
You don't re-use these variables. I think you can safely embed them in the boolean expression below